moveit_ros/trajectory_cache/src/utils/utils.cpp

```diff
@@ -0,0 +1,659 @@
+
+        tf2_quat_frame_offset.normalize();
+        tf2_quat_goal_offset.normalize();
+
+        auto final_quat = tf2_quat_goal_offset * tf2_quat_frame_offset;
+        final_quat.normalize();
+
+        metadata.append(meta_name + ".target_point_offset.x", final_quat.getX());
+        metadata.append(meta_name + ".target_point_offset.y", final_quat.getY());
+        metadata.append(meta_name + ".target_point_offset.z", final_quat.getZ());
+        metadata.append(meta_name + ".target_point_offset.w", final_quat.getW());
+      }
+    }
+  }
+
+  return moveit::core::MoveItErrorCode::SUCCESS;
+}
+
+// RobotState. =====================================================================================
+
+moveit::core::MoveItErrorCode
+appendRobotStateJointStateAsFetchQueryWithTolerance(Query& query, const moveit_msgs::msg::RobotState& robot_state,
+                                                    const MoveGroupInterface& move_group, double match_tolerance,
+                                                    const std::string& prefix)
+{
+  // Make ignored members explicit
+
+  if (!robot_state.multi_dof_joint_state.joint_names.empty())
+  {
+    RCLCPP_WARN_STREAM(getLogger(), "Ignoring " << prefix << ".multi_dof_joint_states: Not supported.");
+  }
+  if (!robot_state.attached_collision_objects.empty())
+  {
+    RCLCPP_WARN_STREAM(getLogger(), "Ignoring " << prefix << ".attached_collision_objects: Not supported.");
+  }
+
+  // Begin extraction.
+
+  if (robot_state.is_diff)
+  {
+    // If plan request states that the start_state is_diff, then we need to get
+    // the current state from the move_group.
+
+    // NOTE: methyldragon -
+    //   I think if is_diff is on, the joint states will not be populated in all
+    //   of the motion plan requests? If this isn't the case we might need to
+    //   apply the joint states as offsets as well.
+    //
+    // TODO: Since MoveIt also potentially does another getCurrentState() call
+    //   when planning, there is a chance that the current state in the cache
+    //   differs from the state used in MoveIt's plan.
+    //
+    //   This issue should go away once the class is used within the move group's
+    //   Plan call.
+    moveit::core::RobotStatePtr current_state = move_group.getCurrentState();
+    if (!current_state)
+    {
+      // NOTE: methyldragon -
+      //   Ideally we would restore the original state here and undo our changes, however copy of the metadata is not
+      //   supported.
+      std::stringstream ss;
+      ss << "Skipping " << prefix << " query append: "
+         << "Could not get robot state.";
+      return MoveItErrorCode(MoveItErrorCode::UNABLE_TO_AQUIRE_SENSOR_DATA, ss.str());
+    }
+
+    moveit_msgs::msg::RobotState current_state_msg;
+    robotStateToRobotStateMsg(*current_state, current_state_msg);
+
+    for (size_t i = 0; i < current_state_msg.joint_state.name.size(); i++)
```

Comment:
Use `++i` instead

Reply:
Nice catch, misremembered which increment type to use, oops.